RFCs/FS-1138-Avoid-boxing-equality-and-comparison.md

```diff
@@ -0,0 +1,88 @@
+# F# RFC FS-1138 - Avoid boxing on equality and comparison, default to equivalence relation mode equality
+
+The design suggestion [Avoid boxing on equality and comparison](https://github.com/fsharp/fslang-suggestions/issues/1280) has been marked "approved in principle". (In some indeterminate future)
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/1280)
+- [ ] Approved in principle
+- [ ] Implementation (Not started)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- Discussion (Not started)
+
+# Summary
+
+F#'s existing equality and comparison operators will be hidden behind a `PartialEquivalenceRelationComparison` module, with new equality and comparison operators implementing `IEquatable<'T>` and `IComparable<'T>` lookup by default, and supporting `nan = nan` instead of `nan <> nan`. This offers performance benefits and eliminates the historical design flaw that is `nan <> nan`.
+
+# Motivation
+
+Currently, F#'s equality and comparison operators [cause boxing on equality and comparison for value types](https://github.com/dotnet/fsharp/issues/526). However, an attempt to fix this through the use of `IEquatable<T>` and `IComparable<T>` was [rejected](https://github.com/dotnet/fsharp/pull/9404) citing the breakage for `nan` equality.

```

Comment:
The only systematical deviation is NaN comparison which may apply to System.Half, and other floating point representations that have a NaN value.

Do you have more examples of IEquatable<`T> deviating from obj.Equals?

Reply:
I've not seen anyone on the compiler team state that this only affects `nan` equality. The PR linked says nothing about that, it explicitly states that anyone can make `obj.Equals` and `IEquatable<T>.Equals` have different behavior, and this would be a runtime-breaking change for them. Over the course of 13+ years of F# being a supported enterprise product in VS, there is undoubtedly code in the wild that works perfectly today and will be broken silently when there is an updated F# compiler delivered by their tools.

Not saying this means the feature stops, but I've simply not seen "we need preserve the weird `nan` behavior" as the actual reason for not accepting work like this in the past.

Reply:
Ok, so hypothetically someone somewhere could be broken. Any examples?

Reply:
This chat got a bit stale, but @Happypig375 I think you are asking about where `obj.Equals` can result in different response than `IEquatable.Equals`. 

The simplest example is probably that the default implementation for `obj.Equals` is often reference equality, which would return false for two different instances of the same object (let's say two persons that are have field `name = "Happy Pig"`). Using `IEquatable.Equals(a, b)`, this would return `true` instead.

Another case might be certain time and date instances, perhaps from NodaTime, but I haven't checked, where two instances may be bitwise unequal, but considered equal after resolving, say, the timezone.

However, an argument can be made that if you implement `IEquality.Equals` that you would want it to behave similar to `object.Equals`.

Reply:
In fact, reading more on it in several blog posts and about Microsoft accepting the oddity of `nan = nan => false`, `nan <> nan => false`, but forcing the incompatible change for `nan.Equals(nan) => true` and `(nan :> IEquatable<float>).Equals(nan) => true` strengthens me in the believe that making this change would be a **good thing** ™️. :)